File: agenkit-cpp/src/adapters/echo_agent.cpp
/**
 * @file echo_agent.cpp
 * @brief Implementation of Echo agent
 */

#include "agenkit/adapters/echo_agent.hpp"

namespace agenkit {
namespace adapters {

std::string EchoAgent::name() const {
    return "echo";
}

std::future<core::Result<core::Message, core::AgentError>>
EchoAgent::process(core::Message message) {
    // Echo back the content
    auto response = core::Message::with_text(
        "assistant",
        message.content_as_str()
    );

    // Preserve metadata from input message
    for (const auto& [key, value] : message.metadata().items()) {
        response.with_metadata(key, value);
    }

    // Return ready future with ok result
    return core::make_ready_future(
        core::Result<core::Message, core::AgentError>::ok(std::move(response))
    );
}

std::vector<std::string> EchoAgent::capabilities() const {
    return {"echo", "test"};
}

} // namespace adapters
} // namespace agenkit

File: agenkit-rust/examples/echo_agent.rs
///! Echo agent example.
///!
///! This example demonstrates a simple echo agent that returns
///! the input message back to the sender.
use agenkit::core::{Agent, AgentError, Message};
use async_trait::async_trait;

/// Simple echo agent that returns input unchanged.
struct EchoAgent {
    name: String,
}

impl EchoAgent {
    fn new(name: impl Into<String>) -> Self {
        Self { name: name.into() }
    }
}

#[async_trait]
impl Agent for EchoAgent {
    fn name(&self) -> &str {
        &self.name
    }

    async fn process(&self, message: Message) -> Result<Message, AgentError> {
        // Simply return the message content as assistant response
        Ok(Message::new("assistant", message.content.clone()))
    }

    fn capabilities(&self) -> Vec<String> {
        vec!["echo".to_string()]
    }
}

#[tokio::main]
async fn main() {
    // Create an echo agent
    let agent = EchoAgent::new("echo");

    // Create a test message
    let msg = Message::with_text("user", "Hello, agent!");

    println!("Sending message: {}", msg.content_as_str().unwrap());

    // Process the message
    match agent.process(msg).await {
        Ok(response) => {
            println!("Agent response: {}", response.content_as_str().unwrap());
        }
        Err(e) => {
            eprintln!("Error: {}", e);
        }
    }
}
